### ISSUE TITLE
consider changing timeouts to deadlines

### ISSUE THREAD

**Issue by weissi** [apple/swift-nio#603]
I think deadlines generally work better than timeouts. Why? Imagine for example a caller wants to have something complete in no more than 5 seconds. So for example

    doSomething(timeout: .seconds(5))

If `doSomething` kicks off the operation straight away, it's all good. However `doSomething` might need to use two other sub-systems to complete the operation. Example:

    func doSomething(timeout: Timeout) -> EventLoopFuture<Void> {
        return step1(timeout: timeout).then {
            step2(timeout: ???)
        }
    }

So what do we fill into the question marks? Sure, we could write

    func doSomething(timeout: Timeout) -> EventLoopFuture<Void> {
        let start = Date()
        return step1(timeout: timeout).then {
            let step1Stop = Date()
            let timeTaken = step1Stop - start
            let timeRemaining = timeout - timeTaken
            if timeRemaining > 0 {
                step2(timeout: timeRemaining)
            } else {
                return eventloop.newFailedFuture(error: TimeoutError())
            }
        }
    }

but that's incredibly ugly and error prone. With deadlines however the call would look like

    doSomething(timeout: .now() + .seconds(5))

and `doSomething` would look much nicer:

    func doSomething(deadline: Deadline) -> EventLoopFuture<Void> {
        return step1(deadline: deadline).then {
            step2(deadline: deadline)
        }
    }

and the `stepX` functions would just need to check if they're still within the deadline.

Basically, deadlines just compose much better than timeouts. So I think for NIO 2.0 we should make this change. @normanmaurer / @Lukasa for input.

**Comment by fumoboy007** [comment]
FWIW, Go also uses deadlines. [`context.WithTimeout`](https://golang.org/pkg/context/#WithTimeout) simply wraps `context.WithDeadline`.

**Comment by fumoboy007** [comment]
Also, some protocols such as [TChannel](https://github.com/uber/tchannel/blob/master/docs/protocol.md) use deadlines and the deadline of the inbound request is expected to be reused for outbound requests.

**Comment by AlanQuatermain** [comment]
I think this sounds like a good idea. Generally speaking the caller of an API is expecting it to function in this way, and I've often found myself (as an API implementor) converting the timeout to a hard date as the first thing I do within a function, ready to pass the resulting deadline to other pieces as necessary (and I've hit the corresponding bugs when I failed to do so).

FWIW, I think there's a use for both `timeout` and `deadline` as distinct concepts. A `timeout` would usually be set as a preference—all datagrams created and set by `ThingX` must be ACKed within `timeout` seconds. Meanwhile, when performing an action via API, such as a `connect` or `bind`, I'd want to specify hard it-must-be-complete-by `deadline`. I think the nomenclature could be used well do distinguish between these similar but non-equivalent concepts.

**Comment by weissi** [comment]
@Lukasa  / @normanmaurer is this something for 2.0 or is it more 3.0?

**Comment by Lukasa** [comment]
Happy to do this in 2.0 if we think we have the bandwidth to do the work.

**Comment by weissi** [comment]
ok, 2.0 stretch goal :)

**Comment by normanmaurer** [comment]
+1

**Comment by JovanMilenkovic** [comment]
Should Dispatch be used for this one?

**Comment by weissi** [comment]
Done 👍👍
